File: source/TGPlaylist.m
//
//  TGPlaylist.m
//  Proto3
//
//  Created by Teo Sartori on 01/07/2013.
//  Copyright (c) 2013 Teo Sartori. All rights reserved.
//

#import "TGPlaylist.h"
#import "rediscover-swift.h"

@protocol SongIDProtocol;
@protocol TGSong;

@implementation TGPlaylist

- (id)init {
    self = [super init];
    if (self) {
        songList = [[NSMutableArray alloc] init];
    }
    
    return self;
}


- (NSUInteger)songsInPlaylist {
    return [songList count];
}

- (id<SongIDProtocol>)songIDAtIndex:(NSUInteger)index {
    return [songList objectAtIndex:index];
}
//- (NSNumber *)songIDAtIndex:(NSUInteger)index {
//    return [songList objectAtIndex:index];
//}

- (void)addSong:(id<SongIDProtocol>)aSongID atIndex:(NSUInteger)index {
    [songList insertObject:aSongID atIndex:index];
//    [songList insertObject:[NSNumber numberWithInteger:aSongID] atIndex:index];
    NSLog(@"Added song %@ to playlist which is now %@",aSongID ,songList);
}

- (void)removeSongAtIndex:(NSUInteger)index {
    [songList removeObjectAtIndex:index];
}

- (void)removeSong:(id)aSong {
    [songList removeObject:aSong];
}
//- (void)removeSong:(NSInteger)aSong {
//    [songList removeObject:[NSNumber numberWithInteger:aSong]];
//}

//- (NSInteger)getNextSongIDToPlay {
- (id<SongIDProtocol>)getNextSongIDToPlay {
//    NSInteger sID = -1;
    id sID = nil;
    NSUInteger songCount = [songList count];
    
    if (songCount) {
        if (_posInPlaylist >= songCount)
            _posInPlaylist = 0;
        
        sID = [songList objectAtIndex:_posInPlaylist++];
//        NSNumber *songID = [songList objectAtIndex:_posInPlaylist++];
//        if (songID != nil) {
//            sID = [songID integerValue];
//        }
    }
    
    return sID;
}

- (void)storeWithName:(NSString *)theName {
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentsDirectory = [paths objectAtIndex:0];
    NSString *fileName = [NSString stringWithFormat:@"%@/%@.m3u",documentsDirectory,theName];
    NSLog(@"The playlist file url is :\n%@",fileName);
    NSString *m3uHeader = @"#EXTM3U\n";
    NSString *m3uExtInf = @"#EXTINF:";
    
    // Start the content off with the header.
    NSString *content = m3uHeader;
    // TEO: Do some error checking here.

    for (id<SongIDProtocol> songID in songList) {

        id<TGSong> song = [SongPool songForSongId:songID];
        NSDictionary *songData  = [song metadataDict];//[_delegate songDataForSongID:songID];
        NSInteger sDuration     = [[songData valueForKey:@"Duration"] integerValue];
        NSURL *songURL          = [songData valueForKey:@"SongURL"];
        NSString *url           = [songURL absoluteString];
        NSString *info          = [NSString stringWithFormat:@"%@%ld,%@ - %@\n",m3uExtInf,sDuration,[songData valueForKey:@"Artist"],[songData valueForKey:@"Title"]];

        
        // Append the new song to the existing content.
        content = [NSString stringWithFormat:@"%@%@%@\n",content,info,url];
    }
    NSLog(@"The playlist content is:\n %@",content);
    
    [content writeToFile:fileName
              atomically:NO
                encoding:NSStringEncodingConversionAllowLossy
                   error:nil];

}

@end

File: source/TGPlaylist.swift
//
//  TGPlaylist.swift
//  reDiscover
//
//  Created by teo on 28/01/16.
//  Copyright © 2016 Teo Sartori. All rights reserved.
//
import Cocoa

/**
@interface TGPlaylist : NSObject <NSTableViewDataSource> {
NSMutableArray *songList;
//    NSUInteger posInPlaylist;
}

@property NSUInteger posInPlaylist;
//@property id<TGPlaylistDelegate> delegate;
@property id<TGSongPoolDelegate> delegate;


- (void)addSong:(id<SongId>)aSongID atIndex:(NSUInteger)index;
- (void)removeSongAtIndex:(NSUInteger)index;
- (void)removeSong:(id<SongId>)aSong;
- (id<SongId>)getNextSongIDToPlay;
- (void)storeWithName:(NSString *)theName;
- (NSUInteger)songsInPlaylist;
- (id<SongId>)songIDAtIndex:(NSUInteger)index;
@end
*/

protocol PlaylistDelegate {
    func getSong(_ songId : SongId) -> TGSong?
}

class TGPlaylist : NSObject, NSTableViewDataSource {
    
    /// A list of song ids.
    var songList = [SongId]()
    var positionInPlaylist = 0
    
    //var songPoolAPI : SongPoolAccessProtocol?
    var delegate : PlaylistDelegate?
    
    override init() {
        super.init()
    }
    
    func songsInPlaylist() -> Int {
        return songList.count
    }
    
    func getSongId(atIndex index : Int) -> SongId? {
        
        guard index < songList.count else {
            print("ERROR: TGPlaylist getSongId index is out of bounds.")
            return nil
        }
        return songList[index]
    }
    
    func removeSong(atIndex index : Int) {
        guard index < songList.count else {
            print("ERROR: TGPlaylist getSongId index is out of bounds.")
            return
        }
        songList.remove(at: index)
    }
    
    func removeSong(_ songId : SongId) {
        
//        if let idx = songList.indexOf( { SongId.isEqual($0) }) {
        if let idx = songList.index( where: { songId == $0 }) {
            songList.remove(at: idx)
        }
    }
    
    /** Adding to the song list is not a concurrent task so no need to them queue up.
        (for now)
    */
    func addSong(withId songId : SongId, atIndex index : Int) {
        guard index <= songList.count else {
            print("ERROR: TGPlaylist addSong index is out of bounds.")
            return
        }
        
        songList.insert(songId, at: index)
    }
    
    func getNextSongIdToPlay() -> SongId? {
        let songCount = songList.count
        guard songCount > 0 else { return nil }
        
        /// Increment the position in the playlist and wrap it to the song count.
        positionInPlaylist = (positionInPlaylist + 1) % songCount
       
        return songList[positionInPlaylist]
    }
    
    func store(withName fileName : String) {
        
        let paths = NSSearchPathForDirectoriesInDomains( .documentDirectory,
            .userDomainMask,
            true)
        let documentsDirectory = paths[0]
        let fullPath = "\(documentsDirectory)/\(fileName).m3u"
        print("The playlist file url is: \(fullPath)")
        let m3uHeader = "#EXTM3U\n"
        let m3uExtInf = "#EXTINF:"
        
        /// Build the content.
        /// Start with the header.
        var content = m3uHeader
        
        for songId in songList {
            
            guard let song = delegate?.getSong(songId) else { continue }
            
            let songData = song.metadataDict()

            guard let songDuration = songData["Duration"],
                  let artist = songData["Artist"],
                  let title = songData["Title"] else {
                continue
            }
            
            content += "\(m3uExtInf)\(songDuration),\(artist) - \(title)\n"
        }
        
        print("The playlist content is \(content)")
        try! content.write(toFile: fullPath, atomically: false, encoding: String.Encoding.utf8)
    }
}
